bitcoin/src/bip158.rs

```diff
@@ -42,18 +42,28 @@ use core::cmp::{self, Ordering};
 use core::convert::TryInto;
 use core::fmt::{self, Display, Formatter};
 
-use hashes::{siphash24, Hash};
+use hashes::{sha256d, siphash24, Hash, hash_newtype};
 use internals::write_err;
 
-use crate::blockdata::block::Block;
+use crate::blockdata::block::{Block, BlockHash};
 use crate::blockdata::script::Script;
 use crate::blockdata::transaction::OutPoint;
 use crate::consensus::encode::VarInt;
 use crate::consensus::{Decodable, Encodable};
-use crate::hash_types::{BlockHash, FilterHash, FilterHeader};
 use crate::io;
 use crate::prelude::*;
 
+hash_newtype! {
+    /// Filter hash, as defined in BIP-157
+    pub struct FilterHash(sha256d::Hash);
+
+    /// Filter header, as defined in BIP-157
+    pub struct FilterHeader(sha256d::Hash);
+}
+
+crate::hash_types::impl_hashencode!(FilterHash);
+crate::hash_types::impl_hashencode!(FilterHeader);
```

Comment:
This stuff is done in https://github.com/rust-bitcoin/rust-bitcoin/pull/2178

Reply:
Yes, and it's effectively required for this PR because the fields are private. (Well, I could call `from_raw_hash` everywhere but that's annoying.) We can merge yours first if you like - it could make the changes easier to understand. I will just have to check if our PRs don't disagree about some moves. As a rule of thumb, the types should be placed wherever they are constructed by hashing the bytes.